rfcs/0003-mutations.md

```diff
@@ -0,0 +1,280 @@
+      await MyContract.abi
+      await MyContract.address
+
+      // Modify a state object, which relays updates back
+      // to the subscribed dApp
+      const { mutationState } = context.graph
+      mutationState.addTransaction("tx_hash")
+
+      ...
+    },
+    async setEntityName (_, args, context) {
+      ...
+    }
+  }
+}
+
+// Configuration Setters
+const config = {
+  // These function arguments are passed in by the dApp
+  ethereum (provider) {
+    return new ethers.providers.Web3Provider(provider)
+  },
+  ipfs (provider) {
+    return new IPFS(provider)
+  },
+  customProperty (value) {
+    return value + 2
+  }
+}
+
+export default {
+  resolvers,
+  config
+}
+```
+
+### dApp Integration
+```javascript
+const {
+  createMutations,
+  createMutationsLink,
+  useMutation
+} = require("@graphprotocol/mutations-ts")
+const myMutations = require("mutations-js-module")
+
+const mutations = createMutations({
+  mutations: myMutations,
+  subgraph: "my-subgraph",
+  node: "http://localhost:8080",
+  // Configuration Getters
+  config: {
+    ethereum: async () => {
+      const { ethereum } = (window as any)
+      await ethereum.enable()
+      return ethereum
+    },
+    ipfs: "http://localhost:5001",
+    customProperty: 5
+  }
+})
+
+// Create an Apollo Links
+const mutationLink = createMutationLink({ mutations })
+const queryLink = createHttpLink({
+  uri: "http://localhost:5001/subgraphs/name/my-subgraph"
+})
+
+const link = split(
+  ({ query }) => {
+    const node = getMainDefinition(query);
+    return node.kind === "OperationDefinition" &&
+           node.operation === "mutation"
+  },
+  mutationLink,
+  queryLink
+);
+
+// Create Apollo Client
+const client = new ApolloClient({
+  link,
+  cache: new InMemoryCache()
+})
+
+const CREATE_ENTITY = gql`
+  mutation createEntity($options: MyEntityOptions) {
+    createEntity(options: $options) {
+      id
+      name
+      value
+    }
+  }
+`
+
+// state === resolver's state
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    client,
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+
+// We can also utilize an optimistic response
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    optimisticResponse: {
+      myEntity: {
+        id: "...",
+        name: "...",
+        value: 5,
+      }
+    },
+    update(proxy, { data }) {
+      // result = data.myEntity
+    },
+    onError(error) {
+      ...
+    },
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+```
+
+## Compatibility
+
+No breaking changes will be introduced, as mutations are an optional add-on to a subgraph.
+
+## Drawbacks and Risks
+
+I have some thoughts but they are rather verbose and tangential. Would love some feedback on this from others first.
+
+## Alternatives
+
+The existing alternative that protocol developers are creating for dApp developers has been described above.
+
+## Open Questions
```

Comment:
I think also an important thing to discuss here is that we currently depend on Apollo. 

Also what we talked about adding support for different frameworks, by releasing several framework specific libraries.